diff --git a/web/viewer.html b/web/viewer.html
--- a/web/viewer.html
+++ b/web/viewer.html
@@ -27,11 +27,6 @@ See https://github.com/adobe-type-tools/cmap-resources
     <meta name="google" content="notranslate" />
     <title>PDF.js viewer</title>
 
-  <meta
-      http-equiv="Content-Security-Policy"
-      content="default-src 'none'; script-src 'self' 'wasm-unsafe-eval'; worker-src 'self' blob:; style-src 'self'; img-src 'self' blob: data:; media-src blob:; font-src 'self' data:; connect-src * blob: data:; base-uri 'none'; form-action 'none';"
-    />
-
 <!-- This snippet is used in production (included from viewer.html) -->
 <link rel="resource" type="application/l10n" href="locale/locale.json" />
 <script src="../build/pdf.mjs" type="module"></script>
diff --git a/web/viewer.mjs b/web/viewer.mjs
--- a/web/viewer.mjs
+++ b/web/viewer.mjs
@@ -24091,7 +24091,9 @@ const PDFViewerApplication = {
     const overlayManager = this.overlayManager = new OverlayManager();
     const renderingQueue = this.pdfRenderingQueue = new PDFRenderingQueue();
     renderingQueue.onIdle = this._cleanup.bind(this);
-    const linkService = this.pdfLinkService = new PDFLinkService({
+    const vscode = acquireVsCodeApi();
+    const config = JSON.parse(document.querySelector("#pdf-view-config")?.dataset.config ?? "{}");
+    const linkService = this.pdfLinkService = new VSCodeLinkService(vscode, config.resourceRoot, {
       eventBus,
       externalLinkTarget: AppOptions.get("externalLinkTarget"),
       externalLinkRel: AppOptions.get("externalLinkRel"),
@@ -26381,4 +26383,31 @@ if (document.readyState === "interactive" || document.readyState === "complete") {
 
 export { PDFViewerApplication, AppConstants as PDFViewerApplicationConstants, AppOptions as PDFViewerApplicationOptions };
 
-//# sourceMappingURL=viewer.mjs.map
\ No newline at end of file
+class VSCodeLinkService extends PDFLinkService {
+  #vscode;
+  #resourceRoot;
+
+  constructor(vscode, resourceRoot, ...args) {
+    super(...args);
+    this.#vscode = vscode;
+    this.#resourceRoot = resourceRoot;
+  }
+
+  addLinkAttributes(link, url, newWindow = false) {
+    super.addLinkAttributes(link, url, newWindow);
+    if (
+      this.externalLinkEnabled &&
+      typeof url === "string" &&
+      typeof this.#resourceRoot === "string" &&
+      url.startsWith(this.#resourceRoot)
+    ) {
+      link.onclick = event => {
+        event.preventDefault();
+        this.#vscode.postMessage({ open: url });
+        return false;
+      };
+    }
+  }
+}
+
+//# sourceMappingURL=viewer.mjs.map
