diff --git a/web/viewer.mjs b/web/viewer.mjs
index c3d794a..79c64de 100644
--- a/web/viewer.mjs
+++ b/web/viewer.mjs
@@ -6373,6 +6373,14 @@ class PDFHistory {
       window.history.forward();
     }
   }
+  get canGoBack() {
+    const state = window.history.state;
+    return this._initialized && !this._popStateInProgress && this.#isValidState(state) && state.uid > 0;
+  }
+  get canGoForward() {
+    const state = window.history.state;
+    return this._initialized && !this._popStateInProgress && this.#isValidState(state) && state.uid < this._maxUid;
+  }
   get popStateInProgress() {
     return this._initialized && (this._popStateInProgress || this._blockHashChange > 0);
   }
@@ -6400,6 +6408,7 @@ class PDFHistory {
     if (shouldReplace) {
       window.history.replaceState(newState, "", newUrl);
     } else {
+      this._maxUid = newState.uid;
       window.history.pushState(newState, "", newUrl);
     }
   }
@@ -13347,7 +13347,8 @@ const PDFViewerApplication = {
     const pdfRenderingQueue = new PDFRenderingQueue();
     pdfRenderingQueue.onIdle = this._cleanup.bind(this);
     this.pdfRenderingQueue = pdfRenderingQueue;
-    const pdfLinkService = new PDFLinkService({
+    const vscode = acquireVsCodeApi()
+    const pdfLinkService = new VSCodeLinkService(vscode, {
       eventBus,
       externalLinkTarget: AppOptions.get("externalLinkTarget"),
       externalLinkRel: AppOptions.get("externalLinkRel"),
@@ -13429,7 +13438,7 @@ const PDFViewerApplication = {
       });
       pdfRenderingQueue.setThumbnailViewer(this.pdfThumbnailViewer);
     }
-    if (!this.isViewerEmbedded && !AppOptions.get("disableHistory")) {
+    if (!AppOptions.get("disableHistory")) {
       this.pdfHistory = new PDFHistory({
         linkService: pdfLinkService,
         eventBus
@@ -15362,4 +15363,25 @@ var __webpack_exports__PDFViewerApplicationConstants = __webpack_exports__.PDFVi
 var __webpack_exports__PDFViewerApplicationOptions = __webpack_exports__.PDFViewerApplicationOptions;
 export { __webpack_exports__PDFViewerApplication as PDFViewerApplication, __webpack_exports__PDFViewerApplicationConstants as PDFViewerApplicationConstants, __webpack_exports__PDFViewerApplicationOptions as PDFViewerApplicationOptions };
 
-//# sourceMappingURL=viewer.mjs.map
\ No newline at end of file
+//# sourceMappingURL=viewer.mjs.map
+
+
+class VSCodeLinkService extends PDFLinkService {
+    #vscode
+    constructor(vscode, ...args){
+        super(...args)
+        this.#vscode = vscode
+    }
+
+    addLinkAttributes(link, url, newWindow = false){
+        if(typeof url === "string" && url.startsWith("https://file+.vscode-resource.vscode-cdn.net")){
+            link.onclick = () => {
+                this.#vscode.postMessage({
+                    open: url
+                });
+            };
+        } else {
+            return super.addLinkAttributes(link, url, newWindow)
+        }
+    }
+}
\ No newline at end of file
